### PR TITLE
{Require,Disallow}OneLinePropertyDocComment sniffs

### LINKED ISSUE #292: Possibility to disallow inline doccomments
It would be great if I could disallow inline doccomments (probably in `SlevomatCodingStandard.Commenting.DocCommentSpacing`?):

```php
class Something
{
    /**
     * @var int
     */
    private $ok;

    /** @var int */
    private $invalid;

}
```

Do you plan to implement something like this?

### PR DESCRIPTION
**Pull request by Majkl578**
Two new sniffs:
* RequireOneLinePropertyDocComment: Requires comments with single-line content expressed as one-liners.
* DisallowOneLinePropertyDocComment: Requires comments expressed as multi-liners.

Closes #292.

This is generalized form of [my InlinePropertyVarTypeHint sniff](https://github.com/cdn77/coding-standard/blob/be9a52f7bff5171a3de7fb13145e26d6bd7c6b61/Cdn77CodingStandard/Sniffs/Classes/InlinePropertyVarTypeHintSniff.php). _Haven't tested on real code (yet)._

@janedbal Do you mind testing this for your use case please? :)

cc @carusogabriel @alcaeus

### CODE REVIEW

File: tests/Sniffs/Classes/RequireOneLinePropertyDocCommentSniffTest.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+		self::assertSniffError($report, 12, RequireOneLinePropertyDocCommentSniff::CODE_MULTI_LINE_PROPERTY_COMMENT);
+		self::assertSniffError($report, 18, RequireOneLinePropertyDocCommentSniff::CODE_MULTI_LINE_PROPERTY_COMMENT);
+		self::assertSniffError($report, 24, RequireOneLinePropertyDocCommentSniff::CODE_MULTI_LINE_PROPERTY_COMMENT);
+		self::assertSniffError($report, 29, RequireOneLinePropertyDocCommentSniff::CODE_MULTI_LINE_PROPERTY_COMMENT);
+		self::assertSniffError($report, 35, RequireOneLinePropertyDocCommentSniff::CODE_MULTI_LINE_PROPERTY_COMMENT);
+		self::assertSniffError($report, 41, RequireOneLinePropertyDocCommentSniff::CODE_MULTI_LINE_PROPERTY_COMMENT);
+		self::assertSniffError($report, 49, RequireOneLinePropertyDocCommentSniff::CODE_MULTI_LINE_PROPERTY_COMMENT);
+		self::assertSniffError($report, 53, RequireOneLinePropertyDocCommentSniff::CODE_MULTI_LINE_PROPERTY_COMMENT);
+
+		self::assertAllFixedInFile($report);
```
Comment (Majkl578) [inline tests/Sniffs/Classes/RequireOneLinePropertyDocCommentSniffTest.php]: @kukulich This does not trigger the `$fix` in the sniff itself (it's `false`), what did I miss? :/

File: README.md [hunk 2]
```diff
@@ -161,6 +161,14 @@ Reports usage of superfluous prefix or suffix "Interface" for interfaces.
 
 Reports usage of superfluous suffix "Exception" for exceptions.
 
+#### SlevomatCodingStandard.Classes.DisallowOneLinePropertyDocComment 🔧
```
Comment (kukulich) [inline README.md]: Please move it to `Commenting` directory.

File: README.md [hunk 3]
```diff
@@ -161,6 +161,14 @@ Reports usage of superfluous prefix or suffix "Interface" for interfaces.
...
+
+Require comments with one-line content to be written as multi-liners.
+
+#### SlevomatCodingStandard.Classes.RequireOneLinePropertyDocComment 🔧
+
+Require comments with one-line content to be written as one-liners.
```
Comment (kukulich) [inline README.md]: Typo -> "with multi-line content" :)

File: SlevomatCodingStandard/Helpers/DocCommentHelper.php [hunk 4]
```diff
@@ -82,7 +82,7 @@ public static function findDocCommentOpenToken(\PHP_CodeSniffer\Files\File $code
 			return $pointer;
 		}
 
-		$found = TokenHelper::findPreviousExcluding($codeSnifferFile, [T_WHITESPACE, T_COMMENT, T_PUBLIC, T_PROTECTED, T_PRIVATE, T_FINAL, T_STATIC, T_ABSTRACT, T_CONST, T_CLASS, T_INTERFACE, T_TRAIT], $pointer - 1);
+		$found = TokenHelper::findPreviousExcluding($codeSnifferFile, [T_WHITESPACE, T_COMMENT, T_PUBLIC, T_PROTECTED, T_PRIVATE, T_VAR, T_FINAL, T_STATIC, T_ABSTRACT, T_CONST, T_CLASS, T_INTERFACE, T_TRAIT], $pointer - 1);
```
Comment (kukulich) [inline SlevomatCodingStandard/Helpers/DocCommentHelper.php]: Can you add this to separate commit please?

File: SlevomatCodingStandard/Helpers/TokenHelper.php [hunk 5]
```diff
@@ -229,7 +229,7 @@ public static function findPreviousLocal(\PHP_CodeSniffer\Files\File $phpcsFile,
 	 */
 	public static function findFirstTokenOnNextLine(\PHP_CodeSniffer\Files\File $phpcsFile, int $pointer): ?int
 	{
-		$newLinePointer = self::findNextContent($phpcsFile, T_WHITESPACE, $phpcsFile->eolChar, $pointer);
+		$newLinePointer = self::findNextContent($phpcsFile, [T_WHITESPACE, T_DOC_COMMENT_WHITESPACE], $phpcsFile->eolChar, $pointer);
```
Comment (kukulich) [inline SlevomatCodingStandard/Helpers/TokenHelper.php]: Can you add this to separate commit please?

File: SlevomatCodingStandard/Sniffs/Classes/DisallowOneLinePropertyDocCommentSniff.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+namespace SlevomatCodingStandard\Sniffs\Classes;
+
+use SlevomatCodingStandard\Helpers\DocCommentHelper;
+use SlevomatCodingStandard\Helpers\PropertyHelper;
+use SlevomatCodingStandard\Helpers\TokenHelper;
+
+class DisallowOneLinePropertyDocCommentSniff implements \PHP_CodeSniffer\Sniffs\Sniff
+{
+
+	public const CODE_SINGLE_LINE_PROPERTY_COMMENT = 'SingleLinePropertyComment';
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/Classes/DisallowOneLinePropertyDocCommentSniff.php]: Inconsistent "one" vs "single"

File: SlevomatCodingStandard/Sniffs/Classes/DisallowOneLinePropertyDocCommentSniff.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+		$lineDifference = $tokens[$docCommentEndPointer]['line'] - $tokens[$docCommentStartPointer]['line'];
+
+		// already multi-line
+		if ($lineDifference !== 0) {
+			return;
+		}
+
+		$fix = $phpcsFile->addFixableError(
+			sprintf(
+				'Found single-line comment for property %s, use multi-line comment instead.',
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/Classes/DisallowOneLinePropertyDocCommentSniff.php]: Again.

File: SlevomatCodingStandard/Sniffs/Classes/DisallowOneLinePropertyDocCommentSniff.php [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		if (!$fix) {
+			return;
+		}
+
+		$commentWhitespacePointer = TokenHelper::findPrevious($phpcsFile, [T_WHITESPACE], $docCommentStartPointer);
+		$indent = ($commentWhitespacePointer !== null ? $tokens[$commentWhitespacePointer]['content'] : '') . ' ';
+
+		$phpcsFile->fixer->beginChangeset();
+
+		$phpcsFile->fixer->addContent($docCommentStartPointer, $phpcsFile->eolChar);
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/Classes/DisallowOneLinePropertyDocCommentSniff.php]: You can use `addNewLine`

File: SlevomatCodingStandard/Sniffs/TestCase.php [hunk 9]
```diff
@@ -105,6 +105,7 @@ protected function assertNoSniffError(\PHP_CodeSniffer\Files\File $codeSnifferFi
 	protected function assertAllFixedInFile(\PHP_CodeSniffer\Files\File $codeSnifferFile): void
 	{
 		$codeSnifferFile->fixer->fixFile();
+
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/TestCase.php]: Useless change but I understand how it has happened :)

File: tests/Sniffs/Classes/RequireOneLinePropertyDocCommentSniffTest.php (hunk 1 above)
Comment (kukulich) [reply tests/Sniffs/Classes/RequireOneLinePropertyDocCommentSniffTest.php]: No idea, will debug after the requested changes.

**Comment by kukulich** [comment]
@Majkl578 Please add one more test to both sniffs so code coverage remains 100% :)

File: tests/Sniffs/Classes/RequireOneLinePropertyDocCommentSniffTest.php (hunk 1 above)
Comment (kukulich) [reply tests/Sniffs/Classes/RequireOneLinePropertyDocCommentSniffTest.php]: It's strange that the lines are covered in code coverage.

File: SlevomatCodingStandard/Sniffs/Classes/RequireOneLinePropertyDocCommentSniff.php [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+			),
+			$docCommentStartPointer,
+			self::CODE_MULTI_LINE_PROPERTY_COMMENT
+		);
+
+		if (!$fix) {
+			return;
+		}
+
+		$contentStartPointer = TokenHelper::findNextExcluding($phpcsFile, [T_DOC_COMMENT_WHITESPACE, T_DOC_COMMENT_STAR], $docCommentStartPointer, $docCommentEndPointer);
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/Classes/RequireOneLinePropertyDocCommentSniff.php]: @Majkl578 I think, there's a bug here.

There should be `$docCommentStartPointer + 1`. Currently `$contentStartPointer` cointains `$docCommentStartPointer`.

That's why the fixer doesn't work.

File: README.md (hunk 3 above)
Comment (Majkl578) [reply README.md]: Actually, no, but changing to _single-line_.

Comment (kukulich) [reply README.md]: Oh, I understand now.

**Comment by Majkl578** [comment]
Should be adressed.

**Comment by Majkl578** [comment]
Tests are failing with lowest deps, yet another PHPCS BC break in 3.2.3...

File: README.md [hunk 11]
```diff
@@ -459,6 +459,14 @@ Reports empty comments.
 
 Reports invalid format of inline phpDocs with `@var`.
 
+#### SlevomatCodingStandard.Commenting.RequireOneLinePropertyDocComment 🔧
+
+Require comments with single-line content to be written as one-liners.
+
+#### SlevomatCodingStandard.Classes.DisallowOneLinePropertyDocComment 🔧
```
Comment (janedbal) [inline README.md]: `Classes` > `Commenting`

**Comment by janedbal** [comment]
Works fine for my use-case :) Thanks!

**Comment by Majkl578** [comment]
@kukulich How do you want to handle the failures due PHPCS version?

**Comment by kukulich** [comment]
I've already merged it.

**Comment by Majkl578** [comment]
Ah you cherry-picked it in meantime. :sweat_smile: Right then, up to you. :)

**Comment by kukulich** [comment]
@Majkl578 https://github.com/slevomat/coding-standard/commit/be1945a76b974b10731feb710d6abfd31523b84d

**Comment by kukulich** [comment]
I've have use merge but didn't notice you rebase it in the meantime :)

**Comment by Majkl578** [comment]
Yup, I was scratching my head how it could conflict with master in non-existent file. Anyway, thanks. :)

**Comment by Majkl578** [comment]
Tbh, the fixer for this exceeded my expectations, worked amazingly well: https://github.com/doctrine/doctrine2/commit/bb1fe1d6c9e3b2ab837445a73b0e76e57683fb78. :100:

**Pull request closed without merging**